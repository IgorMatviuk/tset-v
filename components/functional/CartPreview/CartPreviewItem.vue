<template>
  <div class="cart-preview__item">
    <div class="cart-preview__img-wrapper">
      <ImageLoader>
        <picture>
          <source
            :srcset="variant.base_image.thumb_xs | webp"
            type="image/webp"
          />
          <img
            :src="variant.base_image.thumb_xs"
            :alt="variant.name"
            class="cart-preview__img"
          />
        </picture>
      </ImageLoader>
    </div>
    <div class="cart-preview__content-wrapper">
      <div class="cart-preview__content">
        <NuxtLink
          :to="$i18n.path(variant.url_key)"
          :title="variant.name"
          class="cart-preview__name"
        >
          {{ variant.name }}
        </NuxtLink>
        <div class="cart-preview__options">
          <div
            v-for="attribute in attributes"
            :key="attribute.id"
            class="cart-preview__option"
          >
            {{ attribute.name }}:
            {{ attribute.value }}
          </div>
          <div>Количество: {{ item.quantity }}</div>
        </div>
      </div>
      <div class="cart-preview__price-wrapper">
        <template v-if="variant.discount_price">
          <div class="cart-preview__price">
            {{ variant.formatted_discount_price }}
          </div>
          <div class="cart-preview__price cart-preview__price--discount">
            {{ variant.formatted_price }}
          </div>
        </template>
        <div v-else class="cart-preview__price">
          {{ variant.formatted_price }}
        </div>
      </div>
    </div>
    <button class="cart-preview__remove" type="button" @click="removeItem">
      <svg-icon
        name="trash"
        class="cart-preview__icon"
        width="11"
        height="11"
      />
    </button>
  </div>
</template>

<script>
import ImageLoader from '@/components/shared/ImageLoader'

import removeFromCart from '@/mixins/remove-from-cart'
import imageNextGen from '@/mixins/image-next-gen'

export default {
  components: {
    ImageLoader
  },
  mixins: [removeFromCart, imageNextGen],
  props: {
    item: {
      type: Object,
      required: true
    }
  },
  computed: {
    variant() {
      return this.item.product.variants.find(
        (variant) => variant.id === this.item.variant_id
      )
    },
    attributes() {
      return this.variant.attributes.map((attr) => {
        const tAttr = this.item.product.attributes.find(
          (x) => x.code === attr.code
        )
        const tOpt = tAttr.options.find((x) => x.id === attr.value)

        return {
          name: tAttr.name,
          value: tOpt.label
        }
      })
    }
  }
}
</script>

<style lang="scss">
.cart-preview {
  &__item {
    position: relative;
    display: flex;
    width: 100%;
    padding-bottom: 15px;
    border-bottom: 1px solid #dedcdc;

    &:not(:last-child) {
      margin-bottom: 15px;
    }
  }

  &__content-wrapper {
    display: flex;
    flex-grow: 1;
    justify-content: space-between;
  }

  &__content {
    max-width: 135px;
    padding-right: 8px;
  }

  &__img-wrapper {
    position: relative;
    flex-shrink: 0;
    width: 70px;
    height: 90px;
    margin-right: 15px;
  }

  &__img {
    width: 100%;
    height: auto;
  }

  &__name {
    display: block;
    margin-bottom: 10px;
    overflow: hidden;
    font-size: 12px;
    font-weight: 300;
    text-overflow: ellipsis;
    text-transform: uppercase;
    white-space: nowrap;
  }

  &__options {
    font-size: 10px;
    color: #a8a8a8;
  }

  &__price-wrapper {
    flex-shrink: 0;
    text-align: right;
  }

  &__price {
    font-weight: 500;
    color: $primary;

    &--discount {
      font-size: 11px;
      font-weight: normal;
      color: $text-color;
      text-decoration: line-through;
    }
  }

  &__remove {
    @include reset-button-styles();

    position: absolute;
    right: 0;
    bottom: 0;
    display: flex;
    align-items: center;
    justify-content: flex-end;
    width: 35px;
    height: 35px;
    color: $gray-dark-1;
    transition: color 0.3s;

    @include hover-focus-active() {
      color: $primary;
    }
  }

  &__icon {
    width: 11px;
    height: 11px;
  }
}
</style>
