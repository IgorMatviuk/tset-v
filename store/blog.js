export const mutationTypes = {
  SET_LOADING_START: 'SET_LOADING_START',
  SET_LOADING_SUCCESS: 'SET_LOADING_SUCCESS',
  SET_CURRENT_PAGE: 'SET_CURRENT_PAGE'
}

export const actionsTypes = {
  GET_BLOG_POSTS: 'GET_BLOG_POSTS'
}

export const state = () => ({
  posts: [],
  isLoading: false,
  currentPage: null,
  totalPages: null
})

export const mutations = {
  [mutationTypes.SET_LOADING_START](state, status) {
    state.isLoading = status
  },
  [mutationTypes.SET_CURRENT_PAGE](state, currentPage) {
    state.currentPage = currentPage
  },
  [mutationTypes.SET_LOADING_SUCCESS](
    state,
    { posts, currentPage, totalPages }
  ) {
    state.posts = posts
    state.currentPage = currentPage
    state.totalPages = totalPages

    if (state.isLoading) {
      state.isLoading = false
    }
  }
}

export const actions = {
  async [actionsTypes.GET_BLOG_POSTS]({ commit }, params) {
    commit(mutationTypes.SET_LOADING_START, true)

    let query = Object.keys(params)
      .map((key) => `${key}=${encodeURI(params[key])}`)
      .join('&')

    if (!params.limit) {
      query += '&limit=8'
    }

    const { data, meta } = await this.$axios.$get(`api/posts?${query}`)

    const totalPages = Math.ceil(meta.total / (params.limit || 8))

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      posts: data,
      currentPage: meta.current_page,
      totalPages
    })
  }
}
