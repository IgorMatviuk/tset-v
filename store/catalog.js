import { CATALOG_ITEMS_PER_PAGE } from '@/utils/constants'

function createRootCategories(categories) {
  const withRootList = categories.map((category) => {
    if (category.name_alias) {
      const rootCategory = {
        id: category.id,
        name: category.name_alias,
        slug: category.slug
      }

      category.categories = [
        rootCategory,
        ...createRootCategories(category.categories)
      ]
    }

    return category
  })

  return withRootList
}

export const mutationTypes = {
  SET_CATEGORIES: 'SET_CATEGORIES',
  SET_ACTIVE_CATEGORY: 'SET_ACTIVE_CATEGORY',
  SET_PRODUCTS: 'SET_PRODUCTS',
  SET_ATTRIBUTES: 'SET_ATTRIBUTES',
  SET_FILTERS: 'SET_FILTERS',
  CLEAR_FILTERS: 'CLEAR_FILTERS',
  SET_CATALOG_DATA: 'SET_CATALOG_DATA',
  SET_ORDER: 'SET_ORDER',
  SET_SHOW_ITEMS_COUNT: 'SET_SHOW_ITEMS_COUNT',
  SET_LOADING_TOGGLE: 'SET_LOADING_TOGGLE',
  SET_CURRENT_PAGE: 'SET_CURRENT_PAGE',
  SET_TOTAL_PAGES: 'SET_TOTAL_PAGES',
  RESET_META_DATA: 'RESET_META_DATA',
  SET_SEO_TEXT: 'SET_SEO_TEXT'
}

export const actionTypes = {
  GET_CATEGORIES: 'GET_CATEGORIES',
  GET_PRODUCTS: 'GET_PRODUCTS',
  GET_ATTRIBUTES: 'GET_ATTRIBUTES'
}

export const state = () => ({
  categories: [],
  activeCategory: null,
  products: [],
  attributes: [],
  filters: [],
  order: 'popularity',
  showItemsCount: 4,
  isLoading: false,
  currentPage: null,
  totalPages: null,
  seoText: ''
})

export const mutations = {
  [mutationTypes.SET_CATEGORIES](state, categories) {
    state.categories = categories
  },
  [mutationTypes.SET_ACTIVE_CATEGORY](state, activeCategory) {
    state.activeCategory = activeCategory
  },
  [mutationTypes.SET_PRODUCTS](state, products) {
    state.products = products
  },
  [mutationTypes.SET_ATTRIBUTES](state, attributes) {
    state.attributes = attributes
  },
  [mutationTypes.SET_CATALOG_DATA](state, { category, products, attributes }) {
    state.activeCategory = category
    state.products = products
    state.attributes = attributes
  },
  [mutationTypes.SET_ORDER](state, order) {
    state.order = order
  },
  [mutationTypes.SET_SHOW_ITEMS_COUNT](state, showItemsCount) {
    state.showItemsCount = showItemsCount
  },
  [mutationTypes.SET_FILTERS](state, filters) {
    state.filters = filters
  },
  [mutationTypes.CLEAR_FILTERS](state) {
    state.order = 'popularity'
    state.filters = []
  },
  [mutationTypes.SET_LOADING_TOGGLE](state, status) {
    state.isLoading = status
  },
  [mutationTypes.SET_CURRENT_PAGE](state, currentPage) {
    state.currentPage = currentPage
  },
  [mutationTypes.SET_TOTAL_PAGES](state, totalPages) {
    state.totalPages = totalPages
  },
  [mutationTypes.RESET_META_DATA](state) {
    state.filters = []
    state.order = 'popularity'
    state.currentPage = 1
  },
  [mutationTypes.SET_SEO_TEXT](state, seoText) {
    state.seoText = seoText
  }
}

export const actions = {
  async [actionTypes.GET_CATEGORIES]({ commit }) {
    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: categories } = await this.$axios.$get(
      `${STAGE_API_URL}/api/categories`
    )

    // const { data: categories } = await this.$axios.$get('/api/categories')
    const normilizedCategories = createRootCategories(categories)

    commit(mutationTypes.SET_CATEGORIES, normilizedCategories)
  },
  async [actionTypes.GET_PRODUCTS]({ commit }, { url, query }) {
    commit(mutationTypes.SET_LOADING_TOGGLE, true)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const {
      data: { products },
      meta
    } = await this.$axios.$get(
      `${STAGE_API_URL}/api/router?url=${url}${query ? `&${query}` : ''}`
    )

    const totalPages = Math.ceil(meta.total / CATALOG_ITEMS_PER_PAGE)

    commit(mutationTypes.SET_PRODUCTS, products)
    commit(mutationTypes.SET_CURRENT_PAGE, meta.current_page)
    commit(mutationTypes.SET_TOTAL_PAGES, totalPages)
    commit(mutationTypes.SET_LOADING_TOGGLE, false)
  }
}

export const getters = {
  defaultCategorySlug(state) {
    // FIXME: temporary solution
    return state.categories[0].slug
  },
  optionLabel(state) {
    const { attributes } = state

    return (code, slug) => {
      const { options } = attributes.find((attr) => attr.code === code)
      const { label } = options.find((option) => option.slug === slug)
      return label
    }
  },
  selectedAttributeOptions(state) {
    const { filters } = state

    return (code) => {
      const filter = filters.find((filter) => filter.code === code)
      return filter ? filter.value : null
    }
  }
}
