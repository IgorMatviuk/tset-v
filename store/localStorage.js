export const mutationTypes = {
  SET_CART_ID: 'SET_CART_ID'
}

export const state = () => ({
  cartId: null,
  status: false
})

export const mutations = {
  [mutationTypes.SET_CART_ID](state, cartId) {
    state.cartId = cartId
  }
}
