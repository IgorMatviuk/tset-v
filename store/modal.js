export const mutationTypes = {
  SET_MODAL_META: 'SET_MODAL_META',
  SET_MODAL_VISIBILITY_TOGGLE: 'SET_MODAL_VISIBILITY_TOGGLE',
  SET_MODAL_DATA: 'SET_MODAL_DATA',
  RESET_MODAL_STATE: 'RESET_MODAL_STATE'
}

function createInitialState() {
  return {
    form: {
      visible: false,
      meta: {
        name: '',
        title: '',
        fields: [],
        actionText: '',
        submitText: '',
        links: [],
        action: {}
      },
      data: {}
    },
    message: {
      visible: false,
      meta: {
        title: '',
        content: ''
      }
    },
    testimonials: {
      visible: false,
      data: {}
    }
  }
}

export const state = () => createInitialState()

export const mutations = {
  [mutationTypes.SET_MODAL_META](state, payload) {
    const { type, ...meta } = payload

    if (type === 'form' && meta.success) {
      state.message.meta = meta.success
    }

    state[type].meta = meta
  },
  [mutationTypes.SET_MODAL_VISIBILITY_TOGGLE](state, { type, status }) {
    state[type].visible = status
  },
  [mutationTypes.SET_MODAL_DATA](state, { type, data }) {
    state[type].data = data
  },
  [mutationTypes.RESET_MODAL_STATE](state) {
    const { form, message, testimonials } = createInitialState()

    state.form = form
    state.message = message
    state.testimonials = testimonials
  }
}
