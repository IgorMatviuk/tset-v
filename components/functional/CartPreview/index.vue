<template>
  <div
    id="cartPopoverSync"
    :class="{ 'functions__shopping-bag--disabled': isDisabled }"
    class="functions__item functions__shopping-bag"
  >
    <svg-icon name="shopping-bag" class="functions__icon" />
    <div class="functions__items-count">{{ itemsQty }}</div>
    <BPopover
      :show.sync="isVisible"
      :disabled.sync="isDisabled"
      target="cartPopoverSync"
      placement="bottom"
    >
      <div class="cart-preview">
        <div class="cart-preview__header">
          <span>{{ $t('pages.cart') }}:</span>
          <span>({{ itemsQty }})</span>
        </div>
        <PerfectScrollbar
          :class="{ 'cart-preview__scroll-container--active': itemsQty > 2 }"
          class="cart-preview__scroll-container"
        >
          <div class="cart-preview__body">
            <CartPreviewItem
              v-for="(item, index) in cart.items"
              :key="getPreviewItemKey(item.id, index)"
              :item="item"
            />
          </div>
        </PerfectScrollbar>
        <div class="cart-preview__footer">
          <span>{{ $t('cart.total') }}:</span>
          <b>{{ cart.formatted_grand_total }}</b>
        </div>
        <BButton
          :to="$i18n.path('checkout')"
          block
          variant="primary"
          class="cart-preview__btn-checkout"
        >
          {{ $t('checkoutPage.checkout') }}
        </BButton>
      </div>
      <LoadingOverlay v-if="isLoading" />
    </BPopover>
  </div>
</template>

<script>
import { mapState } from 'vuex'

import CartPreviewItem from './CartPreviewItem'
import LoadingOverlay from '@/components/shared/LoadingOverlay'

export default {
  components: {
    CartPreviewItem,
    LoadingOverlay
  },
  data() {
    return {
      isVisible: false,
      isDisabled: true
    }
  },
  computed: {
    ...mapState({
      cart: (state) => state.cart.cartData,
      isLoading: (state) => state.cart.isLoading
    }),
    itemsQty() {
      return this.cart && this.cart.items_qty ? Number(this.cart.items_qty) : 0
    }
  },
  watch: {
    itemsQty: {
      handler(value) {
        if (!this.isDisabled && !value) {
          this.isDisabled = true

          /* istanbul ignore else */
          if (this.isVisible) {
            this.isVisible = false
          }
        } else if (this.isDisabled && value) {
          this.isDisabled = false
        }

        this.$forceUpdate()
      },
      immediate: true
    }
  },
  mounted() {
    document.addEventListener('mouseup', this.hideOnOutSideClick)
    this.$bus.$on('add-to-cart', this.onAddItem)
  },
  beforeDestroy() {
    document.removeEventListener('mouseup', this.hideOnOutSideClick)
    this.$bus.$off('add-to-cart', this.onAddItem)
  },
  methods: {
    getPreviewItemKey(id, index) {
      return `${id}_${index}`
    },
    hideOnOutSideClick({ target }) {
      const isClickInside =
        target.classList.contains('cart-preview__remove') ||
        target.classList.contains('cart-preview__icon')

      /* istanbul ignore else */
      if (this.isVisible && !isClickInside) {
        this.isVisible = false
      }
    },
    onAddItem() {
      this.isVisible = true
    }
  }
}
</script>

<style lang="scss">
.cart-preview {
  padding: 25px 35px 35px 35px;

  &__scroll-container {
    width: 288px;
    max-height: 228px;
    overflow: auto;

    &--active {
      padding-right: 15px;
    }
  }

  &__scroll-container--active &__content {
    max-width: 120px;
  }

  &__header {
    margin-bottom: 15px;
    font-size: 16px;
    font-weight: bold;
  }

  &__footer {
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding-top: 10px;
    margin-bottom: 14px;
    color: #3e3e3e;
  }
}
</style>
