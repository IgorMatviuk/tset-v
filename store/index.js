import { actionTypes as catalogActionTypes } from '@/store/catalog'

export const mutationTypes = {
  SET_LANG: 'SET_LANG',
  SET_MENUS: 'SET_MENUS',
  SET_PAGE_TYPE: 'SET_PAGE_TYPE',
  SET_BREADCRUMB: 'SET_BREADCRUMB',
  SET_CONTACTS: 'SET_CONTACTS',
  SET_GLOBALS: 'SET_GLOBALS'
}

export const actionTypes = {
  GET_MENUS: 'GET_MENUS',
  GET_CONTACTS: 'GET_CONTACTS',
  GET_BANNER: 'GET_BANNER',
  GET_CLIENT_LOCATION: 'GET_CLIENT_LOCATION'
}

export const state = () => {
  return {
    locale: 'ru',
    locales: ['uk', 'ru'],
    fallbackLocale: 'ru',
    pageType: null,
    breadcrumb: [],
    menus: [],
    globals: {
      countryCode: 'UA',
      banner: null
    },
    contacts: {
      phone: '0 800 332 512',
      email: 'office@virna.com.ua',
      schedule: 'с 9:00 - 18:00',
      socialLinks: [
        { name: 'instagram', url: 'https://www.instagram.com/' },
        { name: 'facebook', url: 'https://facebook.com/' },
        { name: 'youtube', url: 'https://www.youtube.com/' }
      ]
    }
  }
}

export const mutations = {
  [mutationTypes.SET_LANG](state, locale) {
    state.locale = locale
  },
  [mutationTypes.SET_MENUS](state, menus) {
    state.menus = menus
  },
  [mutationTypes.SET_PAGE_TYPE](state, pageType) {
    state.pageType = pageType
  },
  [mutationTypes.SET_BREADCRUMB](state, breadcrumb) {
    state.breadcrumb = breadcrumb
  },
  [mutationTypes.SET_CONTACTS](state, contacts) {
    state.contacts = contacts
  },
  [mutationTypes.SET_GLOBALS](state, { key, value }) {
    state.globals[key] = value
  }
}

export const actions = {
  async [actionTypes.GET_MENUS]({ commit }) {
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: menus } = await this.$axios.$get(`${STAGE_API_URL}/api/menus`)
    commit(mutationTypes.SET_MENUS, menus)
  },
  async [actionTypes.GET_CONTACTS]({ commit }) {
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: contacts } = await this.$axios.$get(
      `${STAGE_API_URL}/api/contacts`
    )
    commit(mutationTypes.SET_CONTACTS, contacts)
  },
  async [actionTypes.GET_BANNER]({ commit }) {
    const { data: banner } = await this.$axios.$get('/api/banner')
    const defaultBanner = {
      key: 'banner',
      value: banner
    }

    commit(mutationTypes.SET_GLOBALS, defaultBanner)
  },
  async [actionTypes.GET_CLIENT_LOCATION]({ commit }) {
    const locationString = await this.$axios.$get('https://ip2c.org/s')
    const [, countryCode] = locationString.split(';')
    const defaultCountry = {
      key: 'countryCode',
      value: countryCode
    }

    commit(mutationTypes.SET_GLOBALS, defaultCountry)
  },
  async nuxtServerInit({ dispatch }) {
    await Promise.all([
      dispatch(actionTypes.GET_MENUS),
      dispatch(actionTypes.GET_CONTACTS),
      dispatch(`catalog/${catalogActionTypes.GET_CATEGORIES}`)
    ])
  }
}

/**
 * Get menum by type
 * @param {object} state
 * @param {('header'|'footer'|'catalog'|'promo')} type
 */
function getMenuByType(state, type) {
  const { menus } = state
  return menus.find((menu) => menu.menu_type === type)
}

export const getters = {
  mobileMenu(state) {
    return getMenuByType(state, 'mobile')
  },
  desktopMenu(state) {
    return getMenuByType(state, 'header')
  },
  promoMenu(state) {
    return getMenuByType(state, 'promo')
  },
  topMenu(state) {
    return getMenuByType(state, 'top')
  },
  catalogMenu(state) {
    return getMenuByType(state, 'catalog')
  },
  footerMenu(state) {
    return getMenuByType(state, 'footer')
  }
}
