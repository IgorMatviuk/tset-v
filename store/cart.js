import { mutationTypes as localStorageMutationTypes } from '@/store/localStorage'

export const mutationTypes = {
  SET_LOADING_START: 'SET_LOADING_START',
  SET_LOADING_SUCCESS: 'SET_LOADING_SUCCESS',
  SET_CART_DATA: 'SET_CART_DATA',
  SET_COUNTRIES: 'SET_COUNTRIES',
  SET_REGIONS: 'SET_REGIONS',
  SET_STATES: 'SET_STATES',
  SET_ORDER_SAVE: 'SET_ORDER_SAVE',
  CLEAR_CART: 'CLEAR_CART'
}

export const actionTypes = {
  GET_CHECKOUT_DATA: 'GET_CHECKOUT_DATA',
  GET_CART: 'GET_CART',
  GET_COUNTRIES: 'GET_COUNTRIES',
  GET_REGIONS: 'GET_REGIONS',
  GET_STATES: 'GET_STATES',
  GET_OFFICES: 'GET_OFFICES',
  ADD_ITEM: 'ADD_ITEM',
  UPDATE_ITEM: 'UPDATE_ITEM',
  REMOVE_ITEM: 'REMOVE_ITEM'
}

export const state = () => ({
  cartData: {},
  countries: [],
  regions: [],
  states: [],
  offices: [],
  isOrderSaved: false,
  isLoading: false
})

export const mutations = {
  [mutationTypes.SET_CART_DATA](state, cartData) {
    state.cartData = cartData
  },
  [mutationTypes.SET_LOADING_START](state) {
    state.isLoading = true
  },
  [mutationTypes.SET_LOADING_SUCCESS](state, { key, value }) {
    if (state.isLoading) {
      state.isLoading = false
    }

    state[key] = value
  },
  [mutationTypes.SET_COUNTRIES](state, countries) {
    state.countries = countries
  },
  [mutationTypes.SET_REGIONS](state, regions) {
    state.regions = regions
  },
  [mutationTypes.SET_STATES](state, states) {
    state.states = states
  },
  [mutationTypes.SET_OFFICES](state, offices) {
    state.offices = offices
  },
  [mutationTypes.SET_ORDER_SAVE](state) {
    state.isOrderSaved = true
    state.cartData = {}
  },
  [mutationTypes.CLEAR_CART](state) {
    state.cartData = {}
    state.isOrderSaved = false
  }
}

export const actions = {
  async [actionTypes.GET_CHECKOUT_DATA]({ commit }) {
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: checkoutData } = await this.$axios.$get(
      `${STAGE_API_URL}/api/checkout`
    )

    commit(mutationTypes.SET_COUNTRIES, checkoutData.countries)
    commit(mutationTypes.SET_REGIONS, checkoutData.regions)
  },
  async [actionTypes.GET_CART]({ commit }, cartId) {
    // const { data: cartData } = await this.$axios.$get('/api/checkout/cart')

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    // const STAGE_API_URL = 'http://localhost:7000'
    // const { data: cartData } = await this.$axios.$get(
    //   `${STAGE_API_URL}/api/checkout/cart`
    // )

    const { data: cartData } = await this.$axios.$get(
      `${STAGE_API_URL}/api/checkout/cart/${cartId}`
    )

    if (cartData) {
      commit(mutationTypes.SET_CART_DATA, cartData)
    }
  },
  async [actionTypes.GET_REGIONS]({ commit }, regionId) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: regions } = await this.$axios.$get(
      `${STAGE_API_URL}/api/shipping/regions`
    )

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'regions',
      value: regions
    })
  },
  async [actionTypes.GET_STATES]({ commit }, regionId) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: states } = await this.$axios.$get(
      `${STAGE_API_URL}/api/shipping/states?region_id=${regionId}`
    )

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'states',
      value: states
    })
  },
  async [actionTypes.GET_OFFICES]({ commit }, stateId) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: offices } = await this.$axios.$get(
      `${STAGE_API_URL}/api/shipping/offices?state_id=${stateId}`
    )

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'offices',
      value: offices
    })
  },
  async [actionTypes.ADD_ITEM]({ commit }, { cartId, itemId }) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    // const { data: cartData } = await this.$axios.$post(
    //   `/api/checkout/cart/add/${id}`
    // )

    const STAGE_API_URL = 'https://back.virna.com.ua'
    // const STAGE_API_URL = 'http://localhost:7000'
    const { data: cartData } = await this.$axios.$post(
      `${STAGE_API_URL}/api/checkout/cart/add`,
      {
        cart: cartId,
        variant: itemId
      }
    )

    if (!state.cartData) {
      commit(
        `localStorage/${localStorageMutationTypes.SET_CART_ID}`,
        cartData.id,
        { root: true }
      )
    }

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'cartData',
      value: cartData
    })
  },
  async [actionTypes.UPDATE_ITEM]({ commit }, { cartId, itemId, body }) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: cartData } = await this.$axios.$put(
      `${STAGE_API_URL}/api/checkout/cart/${cartId}/update/${itemId}`,
      body
    )

    // const { data: cartData } = await this.$axios.$put(
    //   `/api/checkout/cart/update/${id}`,
    //   body
    // )

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'cartData',
      value: cartData
    })
  },
  async [actionTypes.REMOVE_ITEM]({ commit }, { cartId, itemId }) {
    commit(mutationTypes.SET_LOADING_START)

    // FIXME: temporary solution
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: cartData } = await this.$axios.$delete(
      `${STAGE_API_URL}/api/checkout/cart/${cartId}/remove-item/${itemId}`
    )

    if (!cartData) {
      commit(`localStorage/${localStorageMutationTypes.SET_CART_ID}`, null, {
        root: true
      })
    }

    commit(mutationTypes.SET_LOADING_SUCCESS, {
      key: 'cartData',
      value: cartData || {}
    })
  }
}
