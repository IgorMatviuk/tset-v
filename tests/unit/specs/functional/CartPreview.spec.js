import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import { BButton } from 'bootstrap-vue'
import PerfectScrollbar from 'vue2-perfect-scrollbar'
import _cloneDeep from 'lodash.clonedeep'

import CartPreview from '@/components/functional/CartPreview'
import CartPreviewItem from '@/components/functional/CartPreview/CartPreviewItem'

import {
  mutations as cartMutations,
  actionTypes as cartActionTypes,
  actions as cartActions
} from '@/store/cart'
import { mutations as localStorageMutations } from '@/store/localStorage'

import productAttrs from '@/api/factories/product'
import createFactory from '@/api/helpers/factory'
import cartRoute from '@/api/routes/checkout/cart'
import apiResponse from '@/tests/unit/mocks/api-response'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(PerfectScrollbar)
localVue.component('BButton', BButton)

describe('CartPreview', () => {
  let store
  let mockAddToCart
  let mockEventBus

  beforeEach(() => {
    const { data: cartData } = apiResponse(cartRoute)

    cartData.items = createFactory(productAttrs, 4).map((product, index) => {
      const cartProduct = _cloneDeep(product)
      const cartItem = {
        product: cartProduct,
        variant_id: cartProduct.variants[0].id,
        quantity: index + 1
      }

      return cartItem
    })

    const total = cartData.items.reduce((total, item) => {
      const variant = item.product.variants.find(
        (x) => x.id === item.variant_id
      )
      const price = variant.discount_price || variant.price
      total += price * item.quantity
      return total
    }, 0)

    const itemsQty = cartData.items.length

    cartData.grand_total = total.toFixed(4)
    cartData.formatted_grand_total = `${total} грн`
    cartData.items_qty = itemsQty.toFixed(4)

    const app = {
      $axios: {
        $post: jest.fn().mockResolvedValue({ data: { data: cartData } })
      }
    }

    store = new Vuex.Store({
      modules: {
        cart: {
          namespaced: true,
          state: { cartData, isLoading: false },
          mutations: { ...cartMutations },
          actions: {
            [cartActionTypes.ADD_ITEM]: cartActions[
              cartActionTypes.ADD_ITEM
            ].bind(app)
          }
        },
        localStorage: {
          namespaced: true,
          state: { cartId: null },
          mutations: { ...localStorageMutations }
        }
      }
    })

    mockEventBus = {
      $on: jest.fn().mockImplementation((event, callback) => {
        mockAddToCart = jest.fn().mockImplementation((event, payload) => {
          callback(payload)
        })
      }),
      $off: jest.fn(),
      $emit: jest.fn().mockImplementation((...args) => {
        mockAddToCart(...args)
      })
    }
  })

  function createCartPreview(data = {}) {
    const wrapper = mount(CartPreview, {
      localVue,
      store,
      mocks: {
        $t: (str) => str,
        $bus: mockEventBus,
        $i18n: {
          path: jest.fn()
        }
      },
      stubs: ['svg-icon', 'BPopover', 'NuxtLink', 'CartPreviewItem'],
      data() {
        return data
      }
    })

    return wrapper
  }

  test('renders cart items', () => {
    const wrapper = createCartPreview()

    expect(wrapper.findAllComponents(CartPreviewItem).length).toBeGreaterThan(0)
    expect(wrapper.element).toMatchSnapshot()

    wrapper.destroy()
  })

  test('toggles popover visibility', () => {
    const wrapper = createCartPreview()

    const mouseEvent = new MouseEvent('mouseup', { bubbles: true })
    document.documentElement.dispatchEvent(mouseEvent)
    expect(wrapper.vm.isVisible).toBe(false)

    expect(wrapper.vm.isVisible).toBe(false)
    wrapper.setData({ isVisible: true })
    expect(wrapper.vm.isVisible).toBe(true)

    document.documentElement.dispatchEvent(mouseEvent)
    expect(wrapper.vm.isVisible).toBe(false)
  })

  test('shows popover on "add-to-cart" emit', async () => {
    const wrapper = createCartPreview()

    const variantId = store.state.cart.cartData.items[0].variant_id

    expect(wrapper.vm.isVisible).toBe(false)
    store.dispatch(`cart/${cartActionTypes.ADD_ITEM}`, {
      cartId: null,
      itemId: variantId
    })
    wrapper.vm.$bus.$emit('add-to-cart', variantId)

    await wrapper.vm.$nextTick()

    expect(wrapper.vm.isVisible).toBe(true)
    expect(mockAddToCart).toBeCalledWith('add-to-cart', variantId)
  })
})
