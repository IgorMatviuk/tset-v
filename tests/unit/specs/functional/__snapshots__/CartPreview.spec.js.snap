// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPreview renders cart items 1`] = `
<div
  class="functions__item functions__shopping-bag"
  id="cartPopoverSync"
>
  <svg-icon-stub
    class="functions__icon"
    name="shopping-bag"
  />
   
  <div
    class="functions__items-count"
  >
    4
  </div>
   
  <bpopover-stub
    placement="bottom"
    target="cartPopoverSync"
  >
    <div
      class="cart-preview"
    >
      <div
        class="cart-preview__header"
      >
        <span>
          pages.cart:
        </span>
         
        <span>
          (4)
        </span>
      </div>
       
      <div
        class="cart-preview__scroll-container ps cart-preview__scroll-container--active"
      >
        <div
          class="cart-preview__body"
        >
          <cartpreviewitem-stub
            item="[object Object]"
          />
          <cartpreviewitem-stub
            item="[object Object]"
          />
          <cartpreviewitem-stub
            item="[object Object]"
          />
          <cartpreviewitem-stub
            item="[object Object]"
          />
        </div>
        <div
          class="ps__rail-x"
          style="left: 0px; top: 0px;"
        >
          <div
            class="ps__thumb-x"
            style="left: 0px; width: 0px;"
            tabindex="0"
          />
        </div>
        <div
          class="ps__rail-y"
          style="top: 0px; left: 0px;"
        >
          <div
            class="ps__thumb-y"
            style="top: 0px; height: 0px;"
            tabindex="0"
          />
        </div>
      </div>
       
      <div
        class="cart-preview__footer"
      >
        <span>
          cart.total:
        </span>
         
        <b>
          3600 грн
        </b>
      </div>
       
      <button
        class="btn cart-preview__btn-checkout btn-primary btn-block"
        type="button"
      >
        
        checkoutPage.checkout
      
      </button>
    </div>
     
    <!---->
  </bpopover-stub>
</div>
`;
